---
title: "Introduction to survival analysis with R"
author: "Raphaël Bonnet (PhD student C3M-U1065, Inserm, UCA)"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document: default
  word_document: default
  pdf_document: default
tags:
- R
- Survival analysis
- Cox regression model
subtitle: Statistics M1
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction notes

In this course you will learn to work with survival data, and to handle gene expression and annotations matrix.
 
# Access & explore the data

## Loading packages
```{r load packages}
suppressMessages(library(limma))
suppressMessages(library(survival))
suppressMessages(library(survminer))
suppressMessages(library(genefilter))

```

## Loading data

```{r load data}

#load up data

data=readRDS("data_intro.RData")
clinical=data[[1]] #clinical data
dat.m=data[[2]] #data matrix
annot=data[[3]] #probe/gene correspondance
```


## Basic functions R - Exploring the data

* Data matrix are in RNA-DNA seq are constituted of:
    * n rows corresponding to genes called features
    * p columns corresponding to patients called samples 

<br/>


* Before starting to work on a object its always necessary to 
    * know the type of object (character, numeric, boolean, matrix, data.frame, list, S4, etc.)
    * know the dimensions of the object
    
### - Using class(), dim(), colnames(), rownames() and unique() on expression data (dat.m & annot)

```{r dat.m_struct }
#dims and class of dat.m
dim(dat.m)
class(dat.m)

```


```{r names}
# head(), colnames(), rownames()
print('first 5 sample names') 
head(colnames(dat.m))

print('first 5 feature names')
head(rownames(dat.m))

#Bien penser à utliser head() pour n'obtenir qu'un apercu des données
```

We must include gene names into the data matrix

```{r annot }
# Have a look at the annotation table
head(annot)

# ! let's include gene names into the matrix
rownames(dat.m)=annot[,2]
# let's check rownames
head(rownames(dat.m))

#extract gene names
genes=rownames(dat.m)
```

Not all probes have a corresponding annotated genes, this is why we have many genes called NA

### - Using sum() and is.na() on genes (dat.m)

**<span style="color:red">How many genes are NA?</span>** 

**<span style="color:red">How many genes have been annotated?</span>** 


```{r}
print('na genes')
sum(is.na(genes))
print('annotated genes')
sum(!is.na(genes))

#ou nrow(dat.m)-sum(is.na(genes))
```

* Most of the time, and depending on the sequencing library we should find 
    * ~20K protein-coding genes 
    * and ~10K unannnotated pseudogenes (NA)

### - Using class() name() and $ (accessor) on clinical data (clinical)

* Clinical tables in survival studies are constituted of:
    * n rows corresponding to samples
    * p columns corresponding to clinical features 
    
```{r clinical}
class(clinical)
dim(clinical) 

#we can have a look at clinical features
names(clinical)

#two types of clinical data required
unique(clinical$clinical.OS)  #boolean
class(clinical$clinical.OS_months_or_MIN_months_of_OS) #numeric

```

**<span style="color:black"> What are the practical differences between data.frame and matrix objects ? </span>** 

**<span style="color:red"> Check if dimensions matches in dat.m, clinical and annot </span>** 

**<span style="color:red"> What are the number of features, clinical features and samples ? </span>** 


```{r}
#features (annotated)
sum(!is.na(genes))

#samples
nrow(clinical)

#clinical features
ncol(clinical)
```


It's very rare to have access to clinical data, most of the time, researcher must provide the omic data but the clinical infromations remain confidential.


To access them, one must contact the holder of the clinical data, ask for access and wait for authorization and afterwards data. It is best not to rely on that. 

 

## Overview of the gene expression: \ndensity & heatmap & MDS plots


```{r most variable genes init, echo=F}

# lets have a look at the distribution
limma::plotDensities(dat.m,legend=F, main='Features density across samples')
dim(dat.m)
dat.m=varFilter(dat.m, var.cutoff = 0.1)
varGenes <- function(z){y <- apply(z, 1, IQR); z[y > quantile(y, .95),]}

heatmap(varGenes(dat.m), main='Heatmap of the most variable genes')

d <- dist(t(varGenes(dat.m))) # euclidean distances between the rows
fit <- cmdscale(d,eig=TRUE, k=2) # k is the number of dim

# plot solution
x <- fit$points[,1]
y <- fit$points[,2]

colors=clinical$clinical.OS+1
plot(x, y, xlab="mds 1", ylab="mds 2",
  main="Metric MDS", type="n"); text(x, y, labels = colnames(dat.m), cex=.7,col=colors)

url="https://miro.medium.com/max/1478/1*CtT_2Ohfwi-h9kJAVjOYmQ.png"
url2="https://miro.medium.com/proxy/1*KFjJc77t-MGn7-bp3RyYeg.png"
```

> Survival of the patient is overlayed in color (black: remission, red: relapse) 

**These vizualisation gives us some insights on the quality of the expression set we are analysing.**

  * We can have a look to the distribution of the feature (genes) expressions.
  * We can overlook the expression of the top variable genes in a heatmap to see wether or not some patterns already emerge.
  * We can see here that the most variable genes fail to explain the survival of the patients.

<br/>

# Cox Proportional Hazards Regression Analysis

### Basic concepts

<br/>

* __Survival time__: (how long one survive)
    * Event-free survival time: (how long one survive without any event)
    
<br/>

* __Events__ 
    * Relapse 
    * Death
    
<br/>
    
* __Censoring__: end of patient following or incomplete data
    * End of treatment (5 years in general)
    * Loss of consent 
    
<br/>

* __Survival probability__ 

function S(t): survival probability of an individual from t0 to t

S(t) is a **step** function that **changes value only at the time of each event**. 

<br/>

* __Kaplan-Meier plot__

Represent the estimation of the survival probability 
based on true survival times.
*(Kaplan and Meier, J Am Stat Assoc, 1958).*

Provides a useful summary of the data that can be used to estimate measures such as **median survival time**.

<br/>

**<span style="color:blue">Without cencoring in the data</span>**

$$S(t)=(N - n)/N$$

* *Where*
    * *N = the total number of patients* 
    * *n = the number of events from t0 to t*

<br/>


![](`r url`)

<br/>

**<span style="color:red"> Calculate the survival probability at t=7 </span>** 


```{r}
print('survival time t=7')
St7=(10-7)/10
print(St7)
```

<br/>

**<span style="color:blue">With cencoring in the data</span>**
<br/>

$$S(t)=S(t−1)(1−d/n)$$

* *Where*
    * *S(t−1) = the probability of being alive at t−1*
    * *n = the number of patients left just before t*
    * *d = the number of events at t*
    * *dt0 = 0*
    * *S(0) = 1*

<br/>



![](`r url2`)

<br/>

**<span style="color:red"> Calculate the survival probability at t=5 </span>** 

<br/>


```{r}
print('survival time t=5')
St5=0.8*(1-(1/6))
#St4=0.8 d=1 et n=6 car 2 event et 2 censor avant t=5
print(St5)
```

**<span style="color:red"> What are the median survival times for the 2 previous examples? </span>** 

Median survival times are respectively 5 and 6 (here arbitrary unit).

## Implementing the model in R

To fit a cox model to data, we use the following code:

**fit=survfit(Surv(os_time,os) ~ covariate)**

**os_time** is a numeric vector associated to every patients containing the time from diagnosis (t0) to event (death, relapse, cencoring)

**os** is a boolean resulting from the question 'does an event occured?'

**covariate** here set to 1 (no covariate)
  
* __Hazard ratio probalility__

function h(t): event probability at time t

exp(coefficient) gives you the survival ratio between two groups

(i.e: exp(coefficient)=0.2 means that the risk group has 20% more chance of an occuring event)

<br/>


**<span style="color:darkblue">- Using the mean of gene expression of all patients as a cut-off, it is possible to discretize (0: low expression, 1: high expression) the gene expression to create two groups of patients.</span>** 

<br/>


#### - RORC - regulatory role in thymopoiesis

```{r, echo=T}
dat.m=dat.m[,-which(is.na(clinical$clinical.OS_months_or_MIN_months_of_OS))]
clinical=clinical[-which(is.na(clinical$clinical.OS_months_or_MIN_months_of_OS)),]

g=which(rownames(dat.m)=="RORC")
  if (length(g)>1){subset=apply(dat.m[g,],MARGIN = 2,FUN = max)} else {subset=dat.m[g,]}
gene=ifelse(subset>mean(subset),1,0)
RORC=as.numeric(subset)
hist(RORC,breaks = 10)

fit <- survfit(Surv(clinical$clinical.OS_months_or_MIN_months_of_OS,clinical$clinical.OS) ~ gene ,                 
            data = data.frame(clinical))
fit_table <- coxph(Surv(clinical$clinical.OS_months_or_MIN_months_of_OS,clinical$clinical.OS) ~ gene)
info=summary(fit_table)
pval=info$sctest[3]
HR=info$coefficients[2]
if (pval<0.05){print(pval);print(HR)}
fit0 <- survfit(Surv(clinical$clinical.OS_months_or_MIN_months_of_OS,clinical$clinical.OS) ~ 1,
              data = data.frame(clinical))
suppressWarnings(ggsurvplot(list('risk'=fit,'null.model'=fit0),data=clinical,
           pval = TRUE,palette =c("steelblue2", "coral2",'Grey'),legend.labs =
               c("low expression\ngroup","high expression\ngroup","all patients"),linetype = c(1,1,2),censor.shape="x", censor.size = 3,
           title = "\tSurvival model - RORC",xlab = "Time in Month",pval.method =   TRUE,
           ggtheme = theme_bw(), ncensor.plot = F, combine = T,surv.scale=c("percent")))


```

**<span style="color:red">- To what correspond the fit0 variable ?</span>**

fit0 corresponds to the survival of the whole cohort, without covariate that splits data in groups.

**<span style="color:red">- Display the 'fit' variable, what are the proportions of events for each group ?</span>**

```{r}
print(fit)
```

There are 65 patients in first group (gene=0) with 57 events. ~ 87% of events. Mean surv time = 40.2 months

There are 42 patients in first group (gene=1) with 35 events. ~ 87% of events. Mean surv time = 47.7 months

**<span style="color:red">- Display the 'fit_table' variable, have a look at the output</span>**

```{r}
print(fit_table)
```


```{r, echo=F}


url="https://raw.githubusercontent.com/raph06/statistics/master/ranked_survival_score_correspondance.png"
url2="https://raw.githubusercontent.com/raph06/statistics/master/cox_regression_output.png"

```

<br/>

Better survival for low expression group for RORC expression is consistent as it is a prognostic factor.  

**Be very careful when analysing expression data as many other confounding factors can be responsible for the heterogeneity from one cohort to another.**


<br/>

**<span style="color:red">In silico hypothesis must be confronted to a biological experiment</span>**

<br/>


#### Statistics:

![](`r url2`)


**<span style="color:red">- What is the pvalue for this model ?</span>**

Likelihood ratio test p=0.03898
 
**<span style="color:red">- What is the hazard ratio between the two groups ? What does it mean ?</span>**

HR = 0.6380.

As HR < 1 the effect of the covariate is favorable on survival.

## Explore another dataset:


Go to https://stat.ethz.ch/R-manual/R-devel/library/survival/html/lung.html


**<span style="color:red">- How many patients are there ? </span>**

```{r}
#load the data
data(lung)
dim(lung)[1]

```

**<span style="color:red">- How many men - women? </span>**

```{r}
#Number of male, female
male=sum(lung$sex==1)
female=sum(lung$sex==2)

```

**<span style="color:red">- To what correspond the column named status ?</span>**

The status columns corresponds to the survival status of the patients. 1 alive (or censored), 2 dead 

**<span style="color:red">- How many number of events for this variable status ?</span>**

```{r}
#number of events
events=sum(lung$status==2)

```

**<span style="color:red">- How many other covariate are there and to what do they correspond ?</span>**

They are age, sex, ph.ecog, ph.karno, pat.karno, meal.cal wt.loss
See documentation https://stat.ethz.ch/R-manual/R-devel/library/survival/html/lung.html for more info.

**<span style="color:red">- Fit and represent a cox model with no covariate, what is the median survival time?</span>**

```{r}
#fit
 fit0 <- survfit(Surv(time,status) ~ 1 ,                 
            data = data.frame(lung))
```

Mean survival time between 306 and 310.
Use summary(fit0) or print(fit0)

**<span style="color:red">- What is the value of S(t=558) ? </span>**

S(t=558)=23.92%

**<span style="color:red">- Use discrete age variable (cut-off>70) to fit a new model, what do we see ? </span>**
```{r}
#descretize
lung$age=ifelse(lung$age>70,1,0)

fit <- survfit(Surv(time,status) ~ age ,                 
            data = data.frame(lung))
fit_table <- coxph(Surv(clinical$clinical.OS_months_or_MIN_months_of_OS,clinical$clinical.OS) ~ gene)
info=summary(fit_table)
pval=info$sctest[3]
HR=info$coefficients[2]
if (pval<0.05){print(pval);print(HR)}

suppressWarnings(ggsurvplot(list('risk'=fit,'null.model'=fit0),data=lung,
           pval = TRUE,palette =c("steelblue2", "coral2",'Grey'),legend.labs =
               c("low expression\ngroup","high expression\ngroup","all patients"),linetype = c(1,1,2),censor.shape="x", censor.size = 3,
           title = "\tSurvival model - AGE70",xlab = "Time in Days",pval.method =   TRUE,
           ggtheme = theme_bw(), ncensor.plot = F, combine = T,surv.scale=c("percent")))

```







